overlord/configstate/configcore/handlers.go

```diff
@@ -169,3 +191,68 @@ func filesystemOnlyApply(rootDir string, cfg config.ConfGetter, opts *sysconfig.
 	}
 	return nil
 }
+
+// preinstallFilesystemOnlyApply is like filesystemOnlyApply, except that it is
+// only meant to be called before the installation happens, when the image is
+// constructed. This enables certain settings like the pi-config options to
+// be applied to things in the image that must happen before booting, typically
+// bootloader related settings that cannot be applied at runtime without a reboot.
+func preinstallFilesystemOnlyApply(rootDir string, values map[string]interface{}, opts *sysconfig.FilesystemOnlyApplyOptions) error {
+	if rootDir == "" {
+		return fmt.Errorf("internal error: root directory for configcore.FilesystemOnlyApply() not set")
+	}
+
+	if opts == nil {
+		opts = &sysconfig.FilesystemOnlyApplyOptions{}
+	}
+
+	// filter the values to only keep values that use the preinstall flag
+	preinstall := make(map[string]interface{})
+
+	// find the handlers which have the preinstall flag set
+	var preinstallHandlers []configHandler
+	for _, h := range handlers {
+		filter := h.flags().preinstallFilter
+		if filter != nil {
+			currentPreInstallConfigNum := len(preinstall)
+			filter(values, preinstall)
+			if len(preinstall) > currentPreInstallConfigNum {
+				// then the filter for this handler added keys to preinstall map
+				preinstallHandlers = append(preinstallHandlers, h)
+			}
+		}
+	}
+
+	// use only the filtered keys that we got in the above loop for the config
+	cfg := plainCoreConfig(preinstall)
+
+	// validate all of the keys that are relevant to preinstallation against
+	// all of the handlers that are relevant to preinstallation
+	for _, h := range preinstallHandlers {
```

Comment:
maybe there's a way to share code with the other filesystemOnlyApply as Early does with Run or some across all of them (putting helpers in corecfg) if possible?

Reply:
I refactored a bit and broke out the refactors to here specifically: https://github.com/snapcore/snapd/pull/10089